snuba/clickhouse/optimize/optimize.py

```diff
@@ -286,48 +289,58 @@ def optimize_partition_runner(
     2. The final cutoff time is reached. In this case, the scheduler will
     raise an exception which would be propagated to the caller.
+
+    Details of execution flow:
+    1. start by reading configured_num_threads from the Snuba Admin runtime config
+    2. dispatches configured_num_threads threads to optimize configured_num_threads partitions (1 partition per thread)
+    3. as soon as one thread finishes, check configured_num_threads from runtime config again
+    4. if configured_num_threads > number of currently active threads, dispatch more threads
     """
-    remaining_partitions = partitions
-    while remaining_partitions:
-        schedule = scheduler.get_next_schedule(remaining_partitions)
-        num_threads = len(schedule.partitions)
-        logger.info(
-            f"Running schedule with cutoff time: "
-            f"{schedule.cutoff_time} with {num_threads} threads"
-        )
-        threads: MutableSequence[threading.Thread] = []
-        for i in range(0, num_threads):
-            threads.append(
-                threading.Thread(
-                    target=optimize_partitions,
-                    args=(
+    scheduler = OptimizeScheduler(default_parallel_threads=default_parallel_threads)
+
+    with ThreadPoolExecutor(max_workers=32) as executor:
+        futures: set[Future[Any]] = set()
+        # jitter start time
+        for jitter in scheduler.get_start_time_jitter():
+            futures.add(executor.submit(_jitter_start_time, jitter))
+
+        partitions_to_optimize = deque(partitions)
+        while partitions_to_optimize:
+            configured_num_threads = get_num_threads(default_parallel_threads)
+            schedule = scheduler.get_next_schedule(partitions_to_optimize)
+            logger.info(
+                f"Running schedule with cutoff time: "
+                f"{schedule.cutoff_time} with {configured_num_threads} threads"
+            )
+
+            while partitions_to_optimize and len(futures) < configured_num_threads:
+                futures.add(
+                    executor.submit(
+                        optimize_partitions,
                         clickhouse,
                         database,
                         table,
-                        schedule.partitions[i],
+                        [partitions_to_optimize.popleft()],
                         schedule.cutoff_time,
                         tracker,
                         clickhouse_host,
-                        schedule.start_time_jitter_minutes[i]
-                        if schedule.start_time_jitter_minutes
-                        else None,
-                    ),
+                    )
                 )
-            )
 
-            threads[i].start()
+            done, futures = concurrent.futures.wait(
+                futures, return_when=concurrent.futures.FIRST_COMPLETED
+            )
+            for future in done:
+                future.result()
```

Comment:
Should you also remove the future from the futures set so that [this](https://github.com/getsentry/snuba/pull/6175/files#diff-fb97ede04634458aaa8ca52ac172447bbd893822cc750d7765f1272f4d1a7a0fR316) while loop continues in the next iteration?

Reply:
My understanding is that the `futures` set is reassigned by `concurrent.futures.FIRRST_COMPLETED` to only contain futures that haven't finished

Reply:
That makes sense. I feel from a code readability perspective the following changes would help

1. Rename futures to pending_futures.
2. Rename done on [this](https://github.com/getsentry/snuba/pull/6175/files#diff-fb97ede04634458aaa8ca52ac172447bbd893822cc750d7765f1272f4d1a7a0fR327) line to completed_futures.